microsoft-edge/webview2/gettingstarted/wpf.md

```diff
@@ -0,0 +1,291 @@
+
+![https](media/wpf-gettingstarted-https.png)
+
+## Step 7 - Communication between host and web content
+
+The host and the web content can also communicate with each other through `postMessage`. The web content running within a WebView can post to the host through `window.chrome.webview.postMessage`, and the message would be handled by any registered `ICoreWebView2WebMessageReceivedEventHandler` on the host. Likewise, the host can message the web content through `CoreWebView2.PostWebMessageAsString` or `CoreWebView2.PostWebMessageAsJSON`, which would be caught by handlers added from `window.chrome.webview.addEventListener`. The communication mechanism allows the web content to utilize native capabilities by passing messages to ask the host to call native APIs.
+
+As an example to understand the mechanism, when the WebView2 navigates to a URI, lets display that URI in the search bar and alert the user of the URI displayed in the WebView
+
+1. Web Messaging is a **CoreWebView2** functionality. Because the initialization of **CoreWebView2** is asynchronous, create an `InitializeAsync` function that awaits [EnsureCoreWebView2Async](../reference/wpf/0-9-494/microsoft-web-webview2-wpf-webview2#ensurecorewebview2async). In **MainWindow.xaml.cs** update your constructor and define an `InitializeAsync` function:
+
+```csharp
+public MainWindow()
+{
+    InitializeComponent();
+    webView.NavigationStarting += EnsureHTTPS;
+    InitializeAsync();
+}
+
+async void InitializeAsync()
+{
+    await webView.EnsureCoreWebView2Async(null);
+}
+```
+
+2. After **CoreWebView2** is initialized, register an event handler to respond to **Web Message Received**. In this example, the host does three things:
+    1. Receives a web message (the URI) from the web content using [TryGetWebMessageAsString](../reference/dotnet/0-9-494/microsoft-web-webview2-core-corewebview2webmessagereceivedeventargs#trygetwebmessageasstring).
+    2. Updates the text in the search bar to the URI
+    3. Returns the message (the URI) back to the web content using [PostWebMessageAsString](../reference/dotnet/0-9-494/microsoft-web-webview2-core-corewebview2#postwebmessageasstring)
+
+In **MainWindow.xaml.cs** update `InitializeAsync` and define `UpdateSearchBar` as follows:
+
+```csharp
+async void InitializeAsync()
+{
+    await webView.EnsureCoreWebView2Async(null);
+    webView.CoreWebView2.WebMessageReceived += UpdateSearchBar;
+}
+
+void UpdateSearchBar(object sender, CoreWebView2WebMessageReceivedEventArgs args)
+{
+    String Uri = args.TryGetWebMessageAsString();
+    searchBar.Text = Uri;
+    webView.CoreWebView2.PostWebMessageAsString(Uri);
+}
+
+```
+
+3. In order for the WebView to send and respond to the web message, after **CoreWebView2** is initialized, the host:
+
+    1. Injects a script to the web content that registers a handler to print message from the host
+    2. Injects a script to the web content that posts the URL to the host
+
```

Comment:
Can you review and clarify these steps please?

Reply:
The code and the numbered list are not in sync. Can you switch #1 and #2 around in the numbered list? Also can you change "injects a script" to "adds code" ?

Reply:
injects script is the write technical terminology for what is going on here - the numbering has already been removed